L1Trigger/L1TTrackMatch/plugins/L1TkFastVertexProducer.cc

```diff
@@ -0,0 +1,385 @@
+
+  htmp = new TH1F("htmp", ";z (cm); Tracks", nbins, xmin, xmax);
+  htmp_weight = new TH1F("htmp_weight", ";z (cm); Tracks", nbins, xmin, xmax);
+
+  produces<TkPrimaryVertexCollection>();
+}
+
+L1TkFastVertexProducer::~L1TkFastVertexProducer() {
+  // do anything here that needs to be done at desctruction time
+  // (e.g. close files, deallocate resources etc.)
+}
+
+//
+// member functions
+//
+
+// ------------ method called to produce the data  ------------
+void L1TkFastVertexProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
+  using namespace edm;
+
+  std::unique_ptr<TkPrimaryVertexCollection> result(new TkPrimaryVertexCollection);
+
+  // Tracker Topology
+  edm::ESHandle<TrackerTopology> tTopoHandle_;
+  iSetup.get<TrackerTopologyRcd>().get(tTopoHandle_);
+  const TrackerTopology* tTopo = tTopoHandle_.product();
+
+  htmp->Reset();
+  htmp_weight->Reset();
+
+  // ----------------------------------------------------------------------
+
+  if (MonteCarloVertex) {
+    // MC info  ... retrieve the zvertex
+    edm::Handle<edm::HepMCProduct> HepMCEvt;
+    iEvent.getByToken(hepmcToken, HepMCEvt);
+
+    edm::Handle<std::vector<reco::GenParticle> > GenParticleHandle;
+    iEvent.getByToken(genparticleToken, GenParticleHandle);
+
+    const double mm = 0.1;
+    float zvtx_gen = -999;
+
+    if (HepMCEvt.isValid()) {
+      // using HepMCEvt
+
+      const HepMC::GenEvent* MCEvt = HepMCEvt->GetEvent();
+      for (HepMC::GenEvent::vertex_const_iterator ivertex = MCEvt->vertices_begin(); ivertex != MCEvt->vertices_end();
+           ++ivertex) {
+        bool hasParentVertex = false;
+
+        // Loop over the parents looking to see if they are coming from a production vertex
+        for (HepMC::GenVertex::particle_iterator iparent = (*ivertex)->particles_begin(HepMC::parents);
+             iparent != (*ivertex)->particles_end(HepMC::parents);
+             ++iparent)
+          if ((*iparent)->production_vertex()) {
+            hasParentVertex = true;
+            break;
+          }
+
+        // Reject those vertices with parent vertices
+        if (hasParentVertex)
+          continue;
+        // Get the position of the vertex
+        HepMC::FourVector pos = (*ivertex)->position();
+        zvtx_gen = pos.z() * mm;
+        break;  // there should be one single primary vertex
+      }         // end loop over gen vertices
+
+    } else if (GenParticleHandle.isValid()) {
+      std::vector<reco::GenParticle>::const_iterator genpartIter;
+      for (genpartIter = GenParticleHandle->begin(); genpartIter != GenParticleHandle->end(); ++genpartIter) {
+        int status = genpartIter->status();
+        if (status != 3)
+          continue;
+        if (genpartIter->numberOfMothers() == 0)
+          continue;  // the incoming hadrons
+        float part_zvertex = genpartIter->vz();
+        zvtx_gen = part_zvertex;
+        break;  //
+      }
+    } else {
+      edm::LogError("L1TkFastVertexProducer")
```

Comment:
should code execution stop? if so, throw an exception

Reply:
done